### ISSUE TITLE
Cannot mock non-virtual "interface" method on abstract class

### ISSUE THREAD

**Issue by drieseng** [devlooped/moq#1293]
We recently upgraded to version *4.18.2* of **Moq** (from a version that dates back to when the animals still spoke), and found a regression with regards to mocking of non-virtual methods that implement an interface.

To reproduce, compile and run the following code:
```cs
using System;
using System.IO;
using Moq;

publi static class Program
{
    private static void Main()
    {
        var streamFactoryMock = new Mock<IStreamFactory>(MockBehavior.Strict);
        var streamMock = new Mock<Stream>(MockBehavior.Strict);

        var seq = new MockSequence();

        _ = streamFactoryMock.InSequence(seq)
                             .Setup(p => p.Create())
                             .Returns(streamMock.Object);
        _ = streamMock.As<IDisposable>()
                      .InSequence(seq)
                      .Setup(m => m.Dispose());

        using (var s = streamFactoryMock.Object.Create())
        {
        }
    }
}

public interface IStreamFactory
{
    Stream Create();
}
```

**Expected result:**

No exception. This is the result you get up to version **4.7.49** of **Moq**.

**Actual result:**

As of version **4.7.58** (up to **4.18.2**) this results in the following exception:
```txt
Moq.MockException: 'Stream.Close() invocation failed with mock behavior Strict.
All invocations on the mock must have a corresponding setup.'
```

**Comment by stakx** [comment]
TL;DR: change your code as follows:

```diff
         _ = streamFactoryMock.InSequence(seq)
                              .Setup(p => p.Create())
-                             .Returns(streamMock.Object);
+                             .Returns(() => streamMock.Object);
         _ = streamMock.As<IDisposable>()
                       .InSequence(seq)
                       .Setup(m => m.Dispose());
```

Explanation: The `Stream` class already implements the `Dispose` method such that it calls `Close`, and that implementation is not overridable. But you don't want to use that implementation (you're setting up `Dispose` to be a no-op after all), and the only way to do so is to have Moq re-implement the `IDisposable` interface. This can be done by means of `mock.As<IDisposable>()`. However, this call needs to happen _before_ you access `mock.Object`. By making the return value computation lazy, you ensure that `As<IDisposable>` runs first.

(The explanation why your code previously worked would take more time. In short, it used to work because Moq would for some time _always_ re-implement _every_ interface of the mocked type. This was done (IIRC) to make mocking Entity Framework contexts easier, however it broke a ton of other, more frequent use cases, so this was changed back around the time of Moq 4.8 or so. So these days, Moq once again only re-implements interfaces when explicitly requested via `mock.As<TInterface>`.)
